### PR TITLE
Improve fileset distribution during the memo file regeneration

### LINKED ISSUE #148: Predictability during memo file regeneration and general evolution
Since direct, real time translation of image data via Bio-Formats was added to OMERO as "OMERO.fs" [^1] in OMERO 5.0 the issue of how to handle "memo files" has become increasingly important. Each time OMERO.server is released, a statement such as this is included in the release notes [^2]:

> Note that the Bio-Formats Memoizer cache will be invalidated on upgrade from earlier OMERO.server versions.

These memo files are essential for the performant operation of OMERO.server and this microservice. For this microservice each and every rendering operation creates a Bio-Formats reader, reads the memo file, performs the operation and then closes the reader. A current and usable memo file is a hard requirement of the current implementation and if one needs to be created it will always be done in real time, on the fly.

Allowing these memo files to be regenerated in real time is risky as full Bio-Formats reader initialization must take place. This can be especially time consuming for multi-image or complex filesets. To combat this, about 5 years ago (#47), we expanded on the image region infrastructure to allow us to regenerate memo files without requiring continuous access to a running OMERO.server. This was expanded further, about 4 years ago (#60), with a control script which includes parallelization and reporting tools.

5 years on, however, some of the same core issues remain:

1. The time required for the regeneration of individual memo files is hard to predict making accurately downtime estimation during an upgrade difficult if not impossible.
2. It is very difficult to ascertain if or why a change in any reader will cause the the current memo files created by that reader to be invalidated. Furthermore, adding a new reader to Bio-Formats invalidates all memo files. Consequently, a very high level of discipline is required at the Bio-Formats level to minimize memo file invalidation. The impact on OMERO is justifiably not always the primary concern when making changes to Bio-Formats that may affect memo files.
    
    The granularity of the readers to which a change may affect is also hard to ascertain and is hard to test for. Thus we act very conservatively and regularly ensure that all memo files for a given OMERO instance are regenerated.
3. Regenerating all memo files requires visiting every file on the system. This can be exceptionally unpredictable in cases where a large amount of data is in place imported from a storage subsystem is (a) a filesystem view on object storage such as that provided by Amazon S3 File Gateway [^3]; (b) storage whose tiering algorithms depend on access time; (c) otherwise performing poorly.

With these issues in mind the current recommendation is to:

* Run memo regeneration offline well in advance of an upgrade
* In accordance with [2] always assume an upgrade will require a full regeneration of memo files
* Consider [2] and [3] against the workflow, file format use, and storage realities of each deployment when deciding if full regeneration will be beneficial or counter productive

However, this does not mean we cannot consider improvements. A few improvements we might be able to make in the short term have been discussed:

1. Current queries that feed data into the memo file regenerator via the parallelization and batching scripts are a reverse chronological list of filesets to be processed. As each fileset is treated equally there is no attempt to batch intelligently which can create situations of poor resource utilization where one batch finishes quickly and one or more batches do not. Can we use the output of `omero fs importtime` (ome/openmicroscopy#5791) to add per-fileset weights that can be used to more accurately balance resource utilization and achieve more uniform batch execution times as a result?
2. Can we use the output of `omero fs importtime` to estimate runtime?

(`omero fs importtime --cache` may need to be worked on to allow for admins to run on behalf of others in a multi-user environment)

Some longer term items are also worth considering, several of which @kkoz, @sbesson, @melissalinkert, and I have discussed at various times over the years:

1. Work to better localize, assess and document impact of reader changes on memo files (this is currently a lot better than it used to be)
2. In the spirit of some client side improvements (ome/omero-web#543) move further or completely towards asynchronous handling of potentially long running, risky tasks like memo file regeneration
3. Separate metadata rich Bio-Formats reader initialization from binary data only initialization and maintain multiple sets of memo files (these can take wildly different amounts of time)
4. Depart from the "include all" Bio-Formats ImageReader use after import

/cc @stick, @atTODO 

[^1]: https://omero.readthedocs.io/en/stable/developers/Server/FS.html
[^2]: https://www.openmicroscopy.org/2022/12/13/omero-5-6-6.html
[^3]: https://aws.amazon.com/storagegateway/file/s3/

### PR DESCRIPTION
**Pull request by sbesson**
Fixes #148 

This implements the strategy suggested in the issue above to improve both the predictability and the performance of the memo file regeneration.

The SQL query generating the list inputs has been updated  with a new column containing the initial reader initialization time per fileset and sort the results using these values in descending order. The initialization times are computed in a very similar fashion tor to `omero fs importtime` by querying and taking the difference between the image creation timestamp and the fileset upload job timestamp. As for `omero fs importtime`, these values reflect the server-side reader initialization at import time and are unaware of any change since maybe it be on the software side like Bio-Formats reader performance improvements or on the storage side like moving the data through a tiered storage. Nevertheless, this should be a good initial proxy measure allowing to classify the different filesets imported in OMERO Plus and significantly help distributing the regeneration tasks in a parallel environments.

The regen-memo-files.sh script has been updated to divide the SQL results in `chunks` as defined in the [split](https://man7.org/linux/man-pages/man1/split.1.html) command. The output of the command should be a number of input files equal to the number of parallel jobs to be started via `parallel`. The splitting is made using round robin distribution so that the total initialization time is as homogenously distributed as possible between the input files.

Finally a few additional cleanup were added to the `regen-memo-files.sh` script to remove CentOS 6 handling and the `--batch-size` option now superseded by the chunk split.

In terms of functional testing, the best environment would be an OMERO instance containing a reasonable number imported filesets with heteregenous initialization times i.e. from small individual fake files to large plates. The execution time of `regen-memo-files.sh` should be compared with and without these changes. The outcome of the memo file regeneration itself (successes/failures/skips) should be unchanged. With these changes, the overall execution time should be reduced and the parallel jobs should roughly start and finish at the same time.

### CODE REVIEW

**Comment by stick** [comment]
> The regen-memo-files.sh script has been updated to divide the SQL results in chunks as defined in the [split](https://man7.org/linux/man-pages/man1/split.1.html) command. The output of the command should be a number of input files equal to the number of parallel jobs to be started via parallel. The splitting is made using round robin distribution so that the total initialization time is as homogenously distributed as possible between the input files.

So if I understand this correctly `--jobs 4` for example results in 4 files with a round-robin distribution of images to process.  Thereby 4 running processes that are processing their list serially.  

Originally we had concerns over garbage collection with long running `java` processes, as well as startup cost of each process, which is why we're not using parallel to start a process for each individual image.

If this is the intent to only have large lists for each processor there's very little utility in using parallel at all since we're not managing a list that's larger than available processes.

**Comment by sbesson** [comment]
> So if I understand this correctly --jobs 4 for example results in 4 files with a round-robin distribution of images to process. Thereby 4 running processes that are processing their list serially.

Yes that's the current implementation

> Originally we had concerns over garbage collection with long running java processes, as well as startup cost of each process, which is why we're not using parallel to start a process for each individual image.

In the [IDR process](https://github.com/IDR/deployment/blob/master/docs/operating-procedures.md#bio-formats-cache-regeneration)  for regenerating memo cache files which is still using a combination of GNU parallel + `omero render test`, starting one process per fileset was certainly found to be a non-starter for the 6M of individual TIFF files from the HPA project. The IDR strategy includes the extra overhead of initializing the OMERO client which does not apply to the micro-service implementation but I concur startup costs are very real.
I had not considered garbage collection and it would be something worth testing.

> If this is the intent to only have large lists for each processor there's very little utility in using parallel at all since we're not managing a list that's larger than available processes.

A possible compromise would be to split the CSV using round-robin distribution into Nx$JOBS input files where the default value of N is either a fixed value or auto-computed from the number of rows in the CSV and the value of $JOBS to ensure that the default input files have a maximum number of entries (assuming we use the existing 500 by default).

**Comment by sbesson** [comment]
Last commits and especially [7e22384](https://github.com/glencoesoftware/omero-ms-image-region/pull/150/commits/7e22384683e9b0600d26fb4eb2e8a1514ff7063b) should address the concerns raised in https://github.com/glencoesoftware/omero-ms-image-region/pull/150#issuecomment-2347187708 around the potential for growing input files.

The utility now splits the original CSV into N * jobs files using round robin distribution. N is computed using the value of `--batch-size`, 500 by default, which is now interpreted as the maximum number of lines in each split file. For instance:

- a CSV with 1000 entries should be split into 4 files of 250 lines with `--jobs 4`
- a CSV with 1000 entries should be split into 8 files of 125 lines with `--jobs 4 --batch-size 200`
- a CSV with 2000 entries should be split into 4 files of 500 lines with `--jobs 4`
- a CSV with 2400 entries should be split into 8 files of 300 lines with `--jobs 4`
- a CSV with 2400 entries should be split into 8 files of 300 lines with `--jobs 8`

**Review by stick — APPROVED** [review]

**Comment by sbesson** [comment]
@chris-allan have you have a chance to have a look at these changes?

**Comment by sbesson** [comment]
The impact of this change was tested in a development instance with 4 CPUs/ 8GB of RAM with a database including 3797 memo files to regenerate (3686 ok, 1 fail, 110 skipped). The memo regeneration utility was executed 3 times in 3 multi-threaded conditions: `--jobs 2`, `--jobs 3` and no job option i.e. 4 CPUs after cleaning the cache between each iteration.

|        	| without #150           	| with #150         	|
|--------	|------------------------	|-------------------	|
| 2 CPUs 	| 47min 6s +/- 59s       	| 41min 1s +/- 26s  	|
| 3 CPUs 	| 45min 27s +/- 1min 19s 	| 32min 21s +/- 43s 	|
| 4 CPUs 	| 48min 7s +/- 2min 8s   	| DNF               	|

DNF means that that the memo regeneration did not run to completion in any of the three execution as the process ran out of memory after ~12min and was terminated by the OOM killer.

In summary, there are 2 conflicting effects:
- the usage of the round-robin distribution allows a better distribution of the computational tasks and improves the overall performance of the regeneration job taking advantage of the multiple threads
- the process are putting the system under heavy memory load and increase the chances of running out of memory for high number of threads

My assumption is that the latter observation is a consequence of the ordering of the filesets to regenerate in reverse order of initialization time. This means the longest (and possibly the most memory and or I/O intensive) regeneration processes are scheduled to be started concurrently at the onset of the memo regeneration process. 

/cc @stick @chris-allan @melissalinkert

**Comment by sbesson** [comment]
Good discussion with @stick and @chris-allan regarding https://github.com/glencoesoftware/omero-ms-image-region/pull/150#issuecomment-2601863674.

As per
https://github.com/glencoesoftware/omero-ms-image-region/blob/aa533d66260291321ad9a073204ce167483c0f56/src/dist/regen-memo-files.sh#L54

The individual memo regeneration processes are memory constrained up to 2G. For 4 concurrent jobs, this means the total memory that might be allocated by Java is 8GB, which is exactly the total available memory of the system. Considering any overhead and other running processes, the fact that the execution leads to an OOM with `--jobs 4` is not a surprise with this particular configuration. Actually, we should treat the fact that the previous regeneration process with the same amount of concurrency did not ran out of memory as luck.

Overall, the outstanding next step is probably to compute and expose the maximum amount of memory that would be allocated by the command and review our internal documentation of the different options and the trade-offs
